Lib/zipimport.py

```diff
@@ -0,0 +1,640 @@
+        if len(raw_data) != data_size:
+            raise OSError("zipimport: can't read data")
+
+    if compress == 0:
+        # data is not compressed
+        return raw_data
+
+    # Decompress with zlib
+    try:
+        decompress = _get_decompress_func()
+    except:
+        raise ZipImportError("can't decompress data; zlib not available")
+    return decompress(raw_data, -15)
+
+
+# Lenient date/time comparison function. The precision of the mtime
+# in the archive is lower than the mtime stored in a .pyc: we
+# must allow a difference of at most one second.
+def _eq_mtime(t1, t2):
+    # dostime only stores even seconds, so be lenient
+    return abs(t1 - t2) <= 1
+
+# Given the contents of a .py[co] file, unmarshal the data
+# and return the code object. Return None if it the magic word doesn't
+# match (we do this instead of raising an exception as we fall back
+# to .py if available and we don't want to mask other errors).
+def _unmarshal_code(pathname, data, mtime):
+    if len(data) < 16:
+        raise ZipImportError('bad pyc data')
+
+    if data[:4] != _bootstrap_external.MAGIC_NUMBER:
+        _bootstrap._verbose_message('{!r} has bad magic', pathname)
+        return None  # signal caller to try alternative
+
+    flags = _unpack_uint32(data[4:8])
+    if flags != 0:
+        # Hash-based pyc. We currently refuse to handle checked hash-based
+        # pycs. We could validate hash-based pycs against the source, but it
+        # seems likely that most people putting hash-based pycs in a zipfile
+        # will use unchecked ones.
+        if (_imp.check_hash_based_pycs != 'never' and
+            (flags != 0x1 or _imp.check_hash_based_pycs == 'always')):
+            return None
+    elif mtime != 0 and not _eq_mtime(_unpack_uint32(data[8:12]), mtime):
+        _bootstrap._verbose_message('{!r} has bad mtime', pathname)
+        return None  # signal caller to try alternative
+
+    # XXX the pyc's size field is ignored; timestamp collisions are probably
+    # unimportant with zip files.
+    code = marshal.loads(data[16:])
+    if not isinstance(code, _code_type):
+        raise TypeError(f'compiled module {pathname!r} is not a code object')
+    return code
+
+_code_type = type(_unmarshal_code.__code__)
+
+
+# Replace any occurrences of '\r\n?' in the input string with '\n'.
+# This converts DOS and Mac line endings to Unix line endings.
+def _normalize_line_endings(source):
+    source = source.replace(b'\r\n', b'\n')
+    source = source.replace(b'\r', b'\n')
+    return source
+
+# Given a string buffer containing Python source code, compile it
+# and return a code object.
+def _compile_source(pathname, source):
+    source = _normalize_line_endings(source)
+    return compile(source, pathname, 'exec', dont_inherit=True)
+
+# Convert the date/time values found in the Zip archive to a value
+# that's compatible with the time stamp stored in .pyc files.
+def _parse_dostime(d, t):
+    return time.mktime((
```

Comment:
This looks like a rather cryptic mass of symbols.  Could you add some comments, possibly with a more vertical layout, so that this makes more sense?